### ISSUE TITLE
Advice user to add Flutter SDK, app project, pub cache, etc to antivirus whitelist

### ISSUE THREAD

**Issue by jmagman** [flutter/flutter#61997]
Antivirus apps have been known to delete/disable/make un-executable different subcomponents of Flutter.  Warn the user when common manifestations of this happen.

- `FileSystemException: Failed to set file modification time` https://github.com/flutter/flutter/issues/47134#issuecomment-617848767
- `FileSystemException: Cannot create file, Access is denied` when copying app template from SDK to new project. https://github.com/flutter/flutter/issues/36746
- `FileSystemException: Deletion failed, The process cannot access the file because it is being used by another process` https://github.com/flutter/flutter/issues/25197
- `CreateProcessW failed 5 Access is denied.` https://github.com/flutter/flutter/issues/23058
- Unable to find/run "adb" https://github.com/flutter/flutter/issues/61763#issuecomment-660909042
https://github.com/flutter/flutter/blob/859fce9010895ce5bcdb913a8373b5b1a194a1fb/packages/flutter_tools/lib/src/android/android_device_discovery.dart#L61
https://github.com/flutter/flutter/blob/859fce9010895ce5bcdb913a8373b5b1a194a1fb/packages/flutter_tools/lib/src/android/android_device_discovery.dart#L64
- Missing executables,  `ArgumentError: Invalid argument(s): Cannot find executable for`, `ProcessException: The system cannot find the file specified`
  - sdkmanager https://github.com/flutter/flutter/issues/56628 https://github.com/flutter/flutter/issues/9954
  - gradlew https://github.com/flutter/flutter/issues/56628 https://github.com/flutter/flutter/issues/53790#issuecomment-647125472
  - dart https://github.com/flutter/flutter/issues/50273#issuecomment-584493105
  - gen_snapshot https://github.com/flutter/flutter/issues/16008

### Suggestions
- Output warning to add path/to/Flutter/sdk and path/to/project to their anti-virus whitelist.
[dart:io has a similar error.](https://github.com/dart-lang/pub/blob/03b7f9a5a27f46bf554f2a3bdef965e7354c69b0/lib/src/io.dart#L358-L360)
- Add note to https://flutter.dev/docs/get-started/install ?  Probably not a blanket "turn off anti-virus" but something like if you have not whitelisted your Flutter SDK and Flutter projects, your anti-virus may cause certain issues.

**Comment by zanderso** [comment]
@timsneath from an email thread:
>There are a few specific underlying problems:
> * Enterprise / corporate blocklists (e.g. bit9/Santa) that cause spurious failures (file X is blocked, folder Y is read-only, etc.)
> * At least on Windows, real-time scanning that occurs immediately after new files have been added can lock files, which is a problem when we touch the same file twice. MS has improved the file locking APIs and that's much less of an issue on Win10, but it remains prevalent on Windows 7 and/or with poor-quality AV solutions
> * Windows SmartScreen is pretty aggressive with executables that are not commonly run -- this is a "first 24 hour" problem until the hash is registered.
>
> Scanners are fast moving towards more heuristic / ML-based models, so allowlists are not the magic bullet. But a few things that are good to do:
> 1. Catching errors like ERROR_ACCESS_DENIED (on Windows) and including a "retry" option can mitigate against the real-time scanning problem, since by the time the user hits retry the scanner has moved on
> 1. Making sure the failure error message in such cases includes a hint to the user that AV scanners should be evaluated as a potential problem source.
> 1. Avoiding 'suspicious behavior' like modifying existing files in quick succession
> 1. Signing executables with a digital certificate, and ensuring that dependencies are also signed
>
> [Advising users during install to add important Flutter paths to antivirus allowlists is a good suggestion], but more for performance reasons. We might consider a page like this that is also linked from access denied fatal errors.

(1) Sounds good, but we should avoid doing [this](https://en.wikipedia.org/wiki/Abort,_Retry,_Fail%3F). I filed : https://github.com/flutter/flutter/issues/62425 : Improve tool behavior when a FS operation receives ERROR_ACCESS_DENIED on windows.

@jonahwilliams Can you comment on tool behavior around (3).
@godofredoc @athomas Do we do any signing of the `dart` executable for Windows?

**Comment by jonahwilliams** [comment]
My concern about 3) is that without any sort of feedback (i.e. a positive test case) we're unlikely to be able to measure progress on adjusting tool behavior, nor keep from regressing. If there is something very specific that the tool is doing that is _also_ easy to change I don't have a problem with it, but I don't want to end up on a wild bantha chase with nothing to show for.

**Comment by timsneath** [comment]
Do we have any telemetry on install failures today? I'm not sure whether this is just a few individuals with edge cases, or a broader problem.

**Comment by zanderso** [comment]
@jonahwilliams Yeah, that makes sense. @timsneath We tend to see a brief spike in crash rate lasting a week or so after a new release. Unless we have more serious problems, those crashes are dominated by file systems issues, and related issues around spawning subprocesses. This is rather coarse grained compared to what we would need to be able to test/measure to motivate drastically changing e.g. file system access patterns.

**Comment by athomas** [comment]
> @godofredoc @athomas Do we do any signing of the `dart` executable for Windows?

Not for standalone Dart releases (and probably not for Flutter either).

**Comment by TahaTesser** [comment]
The Flutter Daemon has terminated is getting terminated from sleep mode with AV  https://github.com/flutter/flutter/issues/62398

**Comment by jmagman** [comment]
https://github.com/flutter/flutter/issues/65200 mentions some files were identifier as ransomware by Bitdefender.

**Comment by mesahach** [comment]
mine was a similar error the problem was an anti virus on my machine did not know about. after i uninstalled it everything start working well

**Comment by matanlurey** [comment]
@bkonyi Thoughts on this one? It's not clear what specifically the _tool_ can/should do.

**Comment by bkonyi** [comment]
I don't think the tool can reasonably do anything. We'd have to maintain a set of exceptions we've associated with A/V issues in the past and check all thrown exceptions within the tool against this set to print a warning. There's a high likelihood that we'd have false positives and flag some exceptions as being an A/V issue when, in reality, it could be something like the user having bad file system permissions.

I think the most we can do is add a note to the getting started documentation warning about possible issues caused by A/V software and encourage developers to whitelist Flutter related executables. I'm not sure how effective that would actually be, because we can't list a finite set of exceptions that are definitely caused by A/V blocking the executables or file system access.

This is probably best left to StackOverflow answers as there's too many possible failure modes to document.
